File: src/main/java/br/ufsc/bridge/platform/validation/form/errors/FormError.java
package br.ufsc.bridge.platform.validation.form.errors;

import java.time.LocalDate;
import java.util.function.BiConsumer;

import br.ufsc.bridge.metafy.MetaField;
import br.ufsc.bridge.metafy.MetaList;
import br.ufsc.bridge.platform.validation.engine.Rule;

public interface FormError extends ValidationError {

	@Override
	boolean isValid();

	boolean isValid(MetaField<?> field);

	void fieldError(MetaField<?> field, String mensagem);

	FormError formError(MetaField<?> field);

	ListError listError(MetaList<?> field);

	<F> FormError check(MetaField<F> field, Rule<F> rule);

	@Deprecated
	void alfaNumerico(MetaField<String> field);

	@Deprecated
	void beforeToday(MetaField<LocalDate> field);

	@Deprecated
	void cep(MetaField<String> field);

	@Deprecated
	void cpf(MetaField<String> field);

	@Deprecated
	void cnpj(MetaField<String> field);

	@Deprecated
	void cns(MetaField<String> field);

	@Deprecated
	void email(MetaField<String> field);

	@Deprecated
	void empty(MetaField<?> field);

	@Deprecated
	void hour(MetaField<String> field);

	@Deprecated
	void length(MetaField<String> field, int length);

	@Deprecated
	void match(MetaField<String> field, String regex);

	@Deprecated
	void maxLength(MetaField<String> field, int length);

	@Deprecated
	<T extends Number> void maxRange(MetaField<T> field, T maxRange);

	@Deprecated
	void minLength(MetaField<String> field, int minLength);

	@Deprecated
	<T extends Number> void minRange(MetaField<T> field, T minRange);

	@Deprecated
	<T extends Number> void range(MetaField<T> field, T minRange, T maxRange);

	@Deprecated
	void required(MetaField<?> field);

	@Deprecated
	void phone(MetaField<String> field);

	<T> FormError forEach(MetaList<T> field, BiConsumer<T, FormError> itemValidator);

}

File: src/main/kotlin/br/ufsc/bridge/platform/validation/form/errors/FormError.kt
package br.ufsc.bridge.platform.validation.form.errors

import br.ufsc.bridge.metafy.MetaField
import br.ufsc.bridge.metafy.MetaList
import br.ufsc.bridge.platform.validation.engine.Rule
import java.util.function.BiConsumer
import kotlin.reflect.KProperty
import kotlin.reflect.KProperty1

interface FormError<T> : ValidationError {

    override val isValid: Boolean

    fun isValid(field: MetaField<*>): Boolean

    fun <R> isValid(property: KProperty1<T, R>): Boolean

    fun fieldError(field: MetaField<*>, mensagem: String?)

    fun fieldError(property: KProperty<*>, mensagem: String?)

    fun error(mensagem: String?)

    fun <E> formError(field: MetaField<E>): FormError<E>

    fun <R> formError(property: KProperty1<T, R>): FormError<R>

    fun <E> listError(field: MetaList<E>): ListError<E>

    fun <R : List<E>?, E> listError(property: KProperty1<T, R>): ListError<E>

    fun <F> check(field: MetaField<F>, rule: Rule<F>): FormError<T>

    fun <R> check(property: KProperty1<T, R>, rule: Rule<R>): FormError<T>

    fun <E> forEach(field: MetaList<E>, itemValidator: BiConsumer<E, FormError<E>>): FormError<T>

    fun <R : List<E>?, E : Any?> forEach(property: KProperty1<T, R>, itemValidator: (E, FormError<E>) -> Unit): FormError<T>

}
